homeassistant/components/starlink/config_flow.py

```diff
@@ -0,0 +1,77 @@
+"""Config flow for Starlink."""
+from __future__ import annotations
+
+from typing import Any
+
+from starlink_grpc import ChannelContext, GrpcError, get_id
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigFlow
+from homeassistant.const import (
+    CONF_NAME,
+    CONF_URL,
+    DATA_RATE_BITS_PER_SECOND,
+    DATA_RATE_GIGABITS_PER_SECOND,
+    DATA_RATE_MEGABITS_PER_SECOND,
+)
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import CONFIG_LINK_SPEED_FORMAT, CONFIG_POLL_RATE, DOMAIN
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME, default="Starlink"): str,
+        vol.Required(CONF_URL, default="192.168.100.1:9200"): str,
+        vol.Required(CONFIG_POLL_RATE, default=1): vol.All(
+            vol.Coerce(int), vol.Range(min=1)
+        ),
+        vol.Required(
+            CONFIG_LINK_SPEED_FORMAT, default=DATA_RATE_MEGABITS_PER_SECOND
+        ): vol.In(
+            [
+                DATA_RATE_BITS_PER_SECOND,
+                DATA_RATE_MEGABITS_PER_SECOND,
+                DATA_RATE_GIGABITS_PER_SECOND,
+            ]
+        ),
+    }
+)
+
+
+class StarlinkConfigFlow(ConfigFlow, domain=DOMAIN):
+    """The configuration flow for a Starlink system."""
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Ask the user for a server address and a name for the system."""
+        errors = {}
+        if user_input is not None:
+            # Input validation
+            url = user_input[CONF_URL]
+            name = user_input[CONF_NAME]
+            poll_rate = user_input[CONFIG_POLL_RATE]
+            uid = await self.get_device_id(url=url)
+            # If everything looks good, create the entry
+            if uid is not None and poll_rate > 0:
+                return self.async_create_entry(
+                    title=name,
+                    data={
+                        CONF_URL: url,
+                        CONFIG_POLL_RATE: poll_rate,
+                        CONFIG_LINK_SPEED_FORMAT: user_input[CONFIG_LINK_SPEED_FORMAT],
+                    },
+                )
+            errors[CONF_URL] = "server_error"
+        return self.async_show_form(
+            step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
+        )
+
+    async def get_device_id(self, url: str) -> str | None:
+        """Get the device UID."""
+        try:
+            context = ChannelContext(target=url)
+            response = await self.hass.async_add_executor_job(lambda: get_id(context))
+            return response
+        except GrpcError:
+            return None
```

Comment:
Maybe just inline this in the config flow above?

Reply:
Ideally this will be used for auto-discovery (since the typical Starlink setup has Dishy on a fixed IP address), I just haven't figured out how to do that yet